src/trace/entry.zig

```diff
@@ -1,245 +1,83 @@
 const std = @import("std");
-const logfmt = @import("logfmt.zig");
-const Level = @import("level.zig").Level;
-const Channel = @import("../sync/channel.zig").Channel;
+const sig = @import("../sig.zig");
+const trace = @import("lib.zig");
+
+const Level = trace.level.Level;
+const ScopedLogger = trace.log.ScopedLogger;
 const AtomicBool = std.atomic.Value(bool);
 
-pub const Entry = union(enum) {
-    channel_print: ChannelPrintEntry,
-    direct_print: DirectPrintEntry,
-    noop,
+pub fn NewEntry(comptime scope: ?[]const u8) type {
+    return Entry(struct {}, scope);
+}
 
-    const Self = @This();
+pub fn Entry(comptime Fields: type, comptime scope: ?[]const u8) type {
+    return struct {
+        logger: ScopedLogger(scope),
+        level: Level,
+        fields: Fields,
 
-    pub fn init(allocator: std.mem.Allocator, channel: *Channel(logfmt.LogMsg), log_level: Level) Self {
-        return .{ .channel_print = ChannelPrintEntry.init(allocator, channel, log_level) };
-    }
+        const Self = @This();
 
-    pub fn deinit(self: Self) void {
-        switch (self) {
-            .noop => {},
-            .channel_print => |impl| impl.deinit(),
-            .direct_print => |impl| impl.deinit(),
+        pub inline fn init(logger: ScopedLogger(scope), level: Level) Self {
```

Comment:
Unjustified usage of `inline`, no reason for it to be inline.

Reply:
I'm fine with removing inline but just want to go through the reasoning a bit more to flesh out any grey area.

> To add to my "removing inline" comments, inline is a very specific keyword that should only be used to 1. propagate comptime-ness in specific functions or 2. there is an obvious and clear benefit from it, either by i.e removing call frames, LLVM failing to inline the call itself, or something else.

In the past, I've actually made the same argument about avoiding the use of the inline keyword. The basic idea is that you should either trust that the compiler is smarter than you are about making this decision, or prove otherwise. This idea makes sense in principle. Generally I assume the compiler knows better than I do. But occasionally I feel pretty confident about some judgment, and there's a cost in proving myself right: it takes time.

In this case, I assume the compiler agrees with me about inlining these. It seems absurd to push an extra stack frame for each of these method calls, especially for the `log` and `logf` methods which do nothing except pass the exact same parameters to another function with the same signature. In the event that the compiler doesn't agree, I'm betting that it's wrong. But I was too lazy to actually investigate it. Maybe I'm being too bold, assuming the compiler is wrong if it disagrees with me. Maybe the compiler always knows best. I just haven't collected enough evidence about this to be 100% convinced of it.

Let's look at this from another angle. Sometimes a programmer is faced with a related decision. Should I write some code directly in function A, or should I extract that code out into function B and call function B from within function A. Assume it makes no difference from a code readability or organizational perspective. Following your line of reasoning, it would always be better to extract the code into a separate function, because it allows the compiler to decide whether to inline it, rather than forcibly inlining it by writing the code directly in function A. Do you agree with that?

Reply:
tagging @InKryption on this since we've had a similar discussion previously

Reply:
>  It seems absurd to push an extra stack frame for each of these method calls, especially for the log and logf methods which do nothing except pass the parameters to another function with the same signature.

Irrelevant, this costs virtually nothing. It's not even a factor that should be considered. No stack frame would be created either way.

> In the event that the compiler doesn't agree, I'm betting that it's wrong. But I was too lazy to actually investigate it. Maybe I'm being too bold, assuming the compiler is wrong if it disagrees with me. Maybe the compiler always knows best. I just haven't collected enough evidence about this to be 100% convinced of it.

There is no "wrong" here, the compiler not inlining will cost us nothing. I am confident that the optimizer will make the correct choice whatever that may be. If you were to go and take extensive benchmarks across all common use cases and really find that adding `inline` were to make it faster, then it could make sense. 

> Following your line of reasoning, it would always be better to extract the code into a separate function, because it allows the compiler to decide whether to inline it, rather than forcibly inlining it by writing the code directly in function A. Do you agree with that?

When code is used multiple times throughout the codebase and this block of code has a clear intention by itself, such as in this case, `field`, which is a component with a clear function, I agree. 
I draw the line at things like getter and setter functions which *are* ridiculous and provide no benefit to just accessing the object directly.

Reply:
My thoughts:

I think there is an argument to be made that, while there's no cost to not inlining the function, there's also no cost to inlining it either in this case. My reasoning: the reason it's not recommended to inline things willy nilly is because it limits what transformations the compiler is able to make to the code; but for this specific function, I am 99.99999% confident there is no transformation the optimizer could feasibly apply beyond simply inlining it (happy to be proven to otherwise). With that in mind, I think it's perfectly reasonable to specify the intent for this function to be inlined (again, specifically in this case, as soon as there's any sort of potential safety check, inner function calls, etc, I think it best to leave it to the optimizer).

I think a bigger issue here is that this `init` function is mostly superfluous, we could remove it by just initializing `Entry` directly. With that done we don't even really have to consider the implications or tradeoffs of inlining this function, because its purpose is already perfectly accomplished by a simpler construct.

Though if a designated initialization really is desired, I'd vote for just having a constant called `init`, which will benefit from the decl literal feature in 0.14.

Reply:
> With that in mind, I think it's perfectly reasonable to specify the intent for this function to be inlined

Our intent is not to inline it. Inlining it provides no change as you've pointed out, and would be distracting and confusing to a reader. "Why was this inlined?" they may ask themselves. 

re init function, I totally agree with all points there, this shouldn't be an inline functions + it'll be nice to use decl literals when we have them.

Reply:
Thanks for all the feedback. I could still see arguments on both sides, but I'm satisfied at this point with removing inline.

Regarding the init function, typically I'll provide an init function if it abstracts away any fields. In this case it abstracts the initialization of the `fields` fields. But it's trivial and only used in one spot, so I'm happy to remove it.

f0c6b320cbe51686c31cf5e6ffffadd2325d52d7